File: Merge Sort/Merge_sort.py
def merge(arr, l, mid, r):
    n1 = mid - l + 1
    n2 = r - mid

    L = [0] * n1
    R = [0] * n2

    for i in range(0, n1):
        L[i] = arr[l + i]

    for i in range(0, n2):
        R[i] = arr[mid + 1 + i]

    i = 0
    j = 0
    k = l

    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1

    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1

def split(arr, l, r):
    if l < r:
        mid = l + (r - l) // 2

        split(arr, l, mid)
        split(arr, mid + 1, r)

        merge(arr, l, mid, r)

if __name

File: Merge Sort/Merge_sort.java
import java.util.Arrays;

public class MergeSort {

    public static void merge(int[] arr, int l, int mid, int r) {
        int n1 = mid - l + 1;
        int n2 = r - mid;

        int[] L = new int[n1];
        int[] R = new int[n2];

        for (int i = 0; i < n1; i++) {
            L[i] = arr[l + i];
        }
        for (int j = 0; j < n2; j++) {
            R[j] = arr[mid + 1 + j];
        }

        int i = 0;
        int j = 0;
        int k = l;  // Start merging from index 'l'

        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }

        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    public static void split(int[] arr, int l, int r) {
        if (l < r) {
            int mid = l + (r - l) / 2;

            split(arr, l, mid);
            split(arr, mid + 1, r);

            merge(arr, l, mid, r);
        }
    }

    public static void main(String[] args) {
        int[] arr = {31, 12, 22, 15, 4, 1, 5, 88, 9};
        int size = arr.length;

        //Printing the unsroted array first
        System.out.print("Unsorted array: ");
        for (int i = 0; i < size; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();

        split(arr, 0, size - 1);
  
        System.out.print("Sorted array: ");
        for (int i = 0; i < size; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();

    }
}
